### PR TITLE
[WIP] descriptor based wallet serialization and import

### PR DESCRIPTION
**Pull request by Sjors**
Introduces a non-backwards compatible wallet type which contains a set of output descriptors.

A new RPC method `importdescriptor` works like `importmulti` but for one descriptor at a time and with fewer options.

Each wallet descriptor has a `purpose`: current receive, current change,  archive (receive & change)

The current receive and change purposes can have one descriptor with address type bech32 and one with address type base58. It builds on top of #15590 for that.

`getnewaddress` finds correct receive descriptor based on address type, which must be either `bech32` or `legacy` (base58).

Usage:
```sh
bitcoin-cli createwallet true true true
bitcoin-cli importdescriptor "wpkh([00000000/84h/1h/0h]tpub.../1/*)#...."
bitcoin-cli importdescriptor "wpkh([00000000/84h/1h/0h]tpub.../1/*)#..." '{"internal": true}, "timestamp": "now"}' 
bitcoin-cli getnewaddress "SegWit"
bitcoin-cli importdescriptor "sh(wpkh([00000000/49h/1h/0h]tpub.../1/*))#...."
bitcoin-cli getnewaddress "Pre SegWit" legacy
bitcoin-cli dumpwallet "dump" # only way to see the descriptors at the moment
```

This initial version will have several limitations (ignoring the TODO list below). It paves the way towards usage with hardware wallets in #14912.

* only available through `createwallet` followed by `importdescriptor`
* no upgrade path yet
* no private keys allowed
* doesn't serialise descriptor cache

TODO:
- [ ] fix methods that deal with labels
- [ ] fix signmessage
- [ ] fix getrawchangeaddress
- [ ] improve getwalletinfo
- [ ] decide on descriptor serialization (currently stores a string)

For followup PR:
* (re)scan wallet transactions matching descriptor
* transaction creation (e.g. get change address from change descriptor)
* private key support
* multisig support

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -800,6 +804,8 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
 
     std::map<uint256, CWalletTx> mapWallet GUARDED_BY(cs_wallet);
 
+    std::map<unsigned int, std::unique_ptr<WalletDescriptor>> m_descriptors GUARDED_BY(cs_wallet);;
```
Comment (practicalswift) [inline src/wallet/wallet.h]: Nit: `;;` at end of line :-)

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -491,6 +490,24 @@ bool CWallet::LoadWatchOnly(const CScript &dest)
     return CCryptoKeyStore::AddWatchOnly(dest);
 }
 
+bool CWallet::LoadDescriptor(const unsigned int nID, std::unique_ptr<WalletDescriptor>descriptor) {
+    m_descriptors[nID] = std::move(descriptor);
+    return true;
+}
+
+bool CWallet::HaveDescriptor(const unsigned int nID) {
```
Comment (practicalswift) [inline src/wallet/wallet.cpp]: Could be `const`?

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -491,6 +490,24 @@ bool CWallet::LoadWatchOnly(const CScript &dest)
...
+    return m_descriptors.find(nID) != m_descriptors.end();
+}
+
+bool CWallet::AddDescriptor(const unsigned int nID, std::string descriptor, int64_t nCreateTime) {
+    auto desc = MakeUnique<WalletDescriptor>(descriptor, nCreateTime);
+    if (!WalletBatch(*database).WriteDescriptor(nID, desc.get())) {
+        return false;
+    }
+    bool res = LoadDescriptor(nID, std::move(desc));
```
Comment (practicalswift) [inline src/wallet/wallet.cpp]: Nit: Just `return LoadDescriptor(nID, std::move(desc));`?

**Comment by sipa** [comment]
@practicalswift There is absolutely no point in giving coding nits on a PR that is just at a proof of concept stage like this.

**Comment by practicalswift** [comment]
@sipa Oh, personally I would be grateful as a PR author if reviewers pointed out areas of improvement (large or small) also in the WIP stage. What do you think @Sjors?

**Comment by practicalswift** [comment]
Perhaps ["Draft pull requests"](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests) could be used to opt out of any code feedback:

> When you create a pull request, you can choose to create a pull request that is ready for review or a draft pull request. […] When you're ready to get feedback on your draft pull request, you can mark your pull request as ready for review.

Either way it would be nice to have something something similar to the developer notes for review work with clearly stated rationales for the guidelines.

**Comment by Sjors** [comment]
I don't mind and usually just fix it in the next update, but whether it's useful depends on the type of nit.

Semi columns, whitespace, consts and return value stuff feedback is not useful imo, because those things are likely to be moved around.

What I find more useful is help like "don't use unique_ptr in this way", or suggestions anywhere I put "TODO: this is terrible".

File: src/wallet/walletdb.cpp [hunk 4]
```diff
@@ -418,6 +424,18 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
                 strErr = "Error reading wallet database: Unknown non-tolerable wallet flags found";
                 return false;
             }
+        } else if (strType == "descriptor") {
+           unsigned int nID;
+           ssKey >> nID;
+           WalletDescriptor desc;
+           ssValue >> desc;
+           // TODO: figure out how to combine ssValue >> with MakeUnique<WalletDescriptor>
```
Comment (Sjors) [inline src/wallet/walletdb.cpp]: @practicalswift e.g. this is the kind of thing I find feedback more useful on in the WIP stage

**Comment by ryanofsky** [comment]
I took a look at this, and it seems well structured. I'm curious what the first use-case will be. Or I guess what's the minimum thing that needs that needs to be implemented here for this to be useful for hardware wallets and #14912? Current state as of df794753284cc1bd991a5cdf0bd747fffb1db3b9 seems to be that a descriptor wallet is a blank wallet that you're allowed to import public key descriptors to, without them being used for anything yet.

It does seem to me that the flag stuff might be overkill. I don't think we need a new flag or changes to createwallet now that we already have blank wallets. It looks like `WALLET_FLAG_DESCRIPTOR_WALLET` just creates a lot of new branches all over the code that don't need to exist or could be replaced with `!m_descriptors.empty()`.

**Comment by Sjors** [comment]
The reason for the flag is to prevent "legacy" keys from getting mixed with descriptors. Maybe later on I discover that such mixing is fine, then I can drop the flag.

**Comment by Sjors** [comment]
It now distinguishes "address source" descriptors (similar to keypool) from "archive" descriptors. There can only be one base58 and one bech32 receive / change address source descriptor.

I implemented `getnewaddress` and `dumpwallet` includes address labels.

I could use feedback on the details of what we're serializing, as well as on how I implemented the serialization / wallet loading code. That's mainly in commits `[wallet] add descriptor (de)serialization
` and `[wallet] descriptor address serialization + derivation`.

I got a little dizzy from tossing around unique pointers and I couldn't figure out how to deserialize a descriptor in one go. The next step is probably to add serializer code to the Descriptor class, but I'm not sure how to do that.

File: src/wallet/walletdb.h [hunk 5]
```diff
@@ -142,6 +145,97 @@ class CKeyMetadata
...
+    {
+        nVersion = WalletDescriptor::CURRENT_VERSION;
+        m_descriptor_string = m_descriptor->ToString();
+    }
+
+    explicit WalletDescriptor(std::string descriptor_string_, int64_t nCreateTime_, uint64_t id_, int purpose_) :
+        m_descriptor_string(descriptor_string_), nCreateTime(nCreateTime_), m_id(id_), m_purpose(purpose_)
+    {
+        nVersion = WalletDescriptor::CURRENT_VERSION;
+        // ReadKeyValue first initalizes an empty WalletDescriptor
```
Comment (practicalswift) [inline src/wallet/walletdb.h]: Should be "initializes" :-)

**Comment by Sjors** [comment]
Closing in favor of @achow101's #15764 which has better momentum.

**Pull request closed without merging**